vms/platformvm/txs/mempool/mempool.go

```diff
@@ -4,42 +4,326 @@
+	m.consumedUTXOs.Put(tx.TxID, tx.InputIDs())
+	m.droppedTxIDs.Evict(tx.TxID)
+	m.currentGas += gasUsed
+
+	m.updateMetrics()
+	m.cond.Broadcast()
+
+	return nil
+}
+
+func (m *Mempool) tryEvictTxs(gasToFree gas.Gas, txToAdd heapTx) error {
+	gasFreed := gas.Gas(0)
+	toEvict := make([]ids.ID, 0)
+
+	for _, tx := range m.minHeap.Iterator() {
+		// Try to evict lower priced txs to make room for the new tx
+		if tx.GasPrice >= txToAdd.GasPrice {
+			continue
+		}
+
+		txID := tx.Tx.TxID
+		toEvict = append(toEvict, txID)
+		gasFreed += tx.gasUsed
+
+		if gasFreed >= gasToFree {
+			break
+		}
+	}
+
+	if gasFreed < gasToFree {
+		return ErrGasCapacityExceeded
+	}
+
+	for _, txID := range toEvict {
+		m.remove(txID)
+	}
+
+	return nil
+}
+
+func (m *Mempool) updateMetrics() {
+	m.numTxsMetric.Set(float64(m.maxHeap.Len()))
+	m.gasAvailableMetric.Set(float64(m.gasCapacity - m.currentGas))
+}
+
+func (m *Mempool) Get(txID ids.ID) (*txs.Tx, bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	tx, ok := m.maxHeap.Get(txID)
+	if !ok {
+		return nil, false
+	}
+
+	return tx.Tx, true
+}
+
+func (m *Mempool) Remove(txID ids.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.remove(txID)
+}
+
+func (m *Mempool) remove(txID ids.ID) {
+	heapTx, ok := m.maxHeap.Remove(txID)
+	if !ok {
+		return
+	}
+
+	m.minHeap.Remove(txID)
+	m.consumedUTXOs.DeleteKey(txID)
+
+	m.currentGas -= heapTx.gasUsed
+
+	m.updateMetrics()
+}
+
+func (m *Mempool) RemoveConflicts(utxos set.Set[ids.ID]) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for _, removed := range m.consumedUTXOs.DeleteOverlapping(utxos) {
+		m.remove(removed.Key)
+	}
+}
+
+func (m *Mempool) Peek() (*txs.Tx, bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	_, tx, ok := m.maxHeap.Peek()
+	return tx.Tx, ok
+}
+
+func (m *Mempool) Iterate(f func(tx *txs.Tx) bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	for _, v := range m.maxHeap.Iterator() {
+		if !f(v.Tx) {
+			return
+		}
+	}
+}
+
+func (m *Mempool) MarkDropped(txID ids.ID, reason error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if _, ok := m.maxHeap.Get(txID); ok {
+		return
+	}
+
+	m.droppedTxIDs.Put(txID, reason)
+}
+
+func (m *Mempool) GetDropReason(txID ids.ID) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	err, _ := m.droppedTxIDs.Get(txID)
+	return err
+}
+
+func (m *Mempool) Len() int {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	return m.maxHeap.Len()
+}
+
+func (m *Mempool) meter(tx txs.UnsignedTx) (gas.Gas, error) {
+	c, err := fee.TxComplexity(tx)
+	if err != nil {
+		return 0, err
+	}
+
+	g, err := c.ToGas(m.weights)
+	if err != nil {
+		return 0, err
+	}
+
+	if g == 0 {
+		return 0, errNoGasUsed
+	}
+
+	return g, nil
+}
+
+// TODO Copy a test from txmempool
+func (m *Mempool) WaitForEvent(ctx context.Context) (common.Message, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for m.maxHeap.Len() == 0 {
+		if err := m.cond.Wait(ctx); err != nil {
+			return 0, err
+		}
 	}
+	return common.PendingTxs, nil
```

Comment:
Can we add a TODO to only return from where when the `maxHeap` has a `GasPrice` `>=` the chain's minimum gas price?

Reply:
Yeah added.